Class modifiers for apis (#4872)
dart-lang/site-www: src/resources/class-modifiers-for-apis.md

there's nothing to worry about and nothing to do.
If the class has a non-factory constructor, an `extends` clause,
or a `with` clause, then it already can't be used as a mixin.
The Dart 3.0 behavior is unchanged.

In practice, this describes about 90% of existing classes.
For the remaining classes that can be used as mixins,
you have to decide what you want to support.

Here are a few questions to help decide, first a pragmatic one, then a couple
of design questions:

*   **Do you want to risk breaking any users?** If the answer is a hard no,
    then simply place `mixin` before any class that could be used as a mixin.
    This exactly preserves the existing behavior of your API.

On the other hand, if you want to take this opportunity to rethink the
affordances your API offers, then you may want to *not* turn it into a `mixin
class`. Consider these two questions:

*   **Do you want users to be able to construct instances of it directly?**
    In other words, is the class not abstract?

*   **Do you *want* people to be able to use the declaration as a mixin?**
    In other words, do you want them to be able to use it in `with` clauses?

> If the answer to both is "yes", then make it a mixin class. If the answer to
> the second is "no", then just leave it as a class. If the answer to the first
> is "no" and the second is "yes", then change it from a class to a mixin
> declaration.

Reviewer: ```suggestion
- If the answer to both is "yes", then make it a mixin class. 
- If the answer to the second is "no", then just leave it as a class. 
- If the answer to the first is "no" and the second is "yes",
  then change it from a class to a mixin declaration.
```

all of the private members your class or mixin declares.
This can help prevent runtime errors that might otherwise occur.

Consider this library:

```dart
// a.dart
class A {
  void _privateMethod() {
    print('I inherited from A');
  }
}

void callPrivateMethod(A a) {
  a._privateMethod();
}
```

This code seems fine on its own,
but there's nothing preventing a user from creating another library like this:

```dart
// b.dart
import 'a.dart';

class B implements A {
  // No implementation of _privateMethod()!
}

main() {
  callPrivateMethod(B()); // Runtime exception!
}
```

Adding the `base` modifier to the class can help prevent these runtime errors.
As with `interface`, you can ignore this restriction
in the same library where the `base` class or mixin is declared,
and then subclasses in the same library
will be reminded to implement the private methods.
But note that the next section *does* apply:

### Base transitivity

The goal of marking a class `base` is to ensure that
every instance of that type concretely inherits from it.
To maintain this, the base restriction is "contagious".
Every subtype of a type marked `base` -- *direct or indirect* --

> must also prevent being implemented too.

Reviewer: "also" or "too", not both
```suggestion
must also prevent being implemented.
```

That means it must be marked `base`
(or `final` or `sealed`, which we'll get to next).

> Applying `base` to a type requires some care then,
> because it affects not just what users can do with your class or mixin,

Reviewer: Shorten sentence
```suggestion
Applying `base` to a type requires some care, then.
It affects not just what users can do with your class or mixin,
```

but also the affordances *their* subclasses can offer.
Once you've put `base` on a type, the whole hierarchy under it
is prohibited from being implemented.

That sounds intense, but it's how most other programming languages
have always worked.
Most don't have implicit interfaces at all,
so when you declare a class in Java, C#, or other languages,

> you effectively have the same limitation.

Reviewer 2: Perhaps call it a 'constraint' rather than a 'limitation': It's actually fair to say that it is a limitation in those other languages, but in the context of Dart it is a deliberate choice: "I want the complete package consisting of some guarantees and some constraints", and if the constraint/limitation part is a problem then just don't write `base`.

If a switch has cases for every direct subtype of a type marked `sealed`,
then the compiler knows the switch is exhaustive.

[exhaustiveness checking]: /language/branches#exhaustiveness-checking

```dart
// amigos.dart
abstract class Amigo {}
class Lucky extends Amigo {}
class Dusty extends Amigo {}
class Ned extends Amigo {}

String lastName(Amigo amigo) =>
    switch (amigo) {
      case Lucky _ => 'Day';
      case Dusty _ => 'Bottoms';
      case Ned _   => 'Nederlander';
    }
```

This switch has a case for each of the subtypes of `Amigo`.
The compiler knows that every instance of `Amigo` must be an instance of one
of those subtypes, so it knows the switch is safely exhaustive and doesn't
require any final default case.

For this to be sound, the compiler enforces two restrictions:

1.  The sealed class can't itself be directly constructible.
    Otherwise, you could have an instance of `Amigo` that isn't
    an instance of *any* of the subtypes.
    So every `sealed` class is implicitly `abstract` too.

2.  Every direct subtype of the sealed type must be in the same library
    where the sealed type is declared.

>     This way the compiler can find them all and know that there aren't

Reviewer: ```suggestion
    This way, the compiler can find them all. It knows that there aren't
```

    other hidden subtypes floating around that would not match any of the cases.

The second restriction is similar to `final`.
Like `final`, it means that a class marked `sealed` can't be directly
extended, implemented, or mixed in outside of the library where it's declared.
But, unlike `base` and `final`, there is no *transitive* restriction:

```dart
// amigo.dart
abstract class Amigo {}
class Lucky extends Amigo {}
class Dusty extends Amigo {}
class Ned extends Amigo {}

// other.dart

// This is an error:
class Bad extends Amigo {}

// But these are both fine:
class OtherLucky extends Lucky {}
class OtherDusty implements Dusty {}
```

Of course, if you *want* the subtypes of your sealed type
to be restricted as well, you can get that by marking them
using `interface`, `base`, `final`, or `sealed`.

> 

Reviewer: I would make this it's own section (it'd be nice to point people to directly). Either top level or a subsection of sealed:

```suggestion

## `sealed` versus `final`

```
